doc/katana.man
.\ manpage for katana
.TH "KATANA" 1 2011-04-13 "Dartmouth College" "Katana Manual"
.SH NAME
.B katana
\- A binary-level hot-patching system for userland and a
general-purpose ELF and DWARF manipulation shell.
.SH SYNOPSIS

.\ disable filling
.nf
Run the Katana shell interactively:
\fBkatana\fP

Run a Katana shell script:
\fBkatana\fP SCRIPT_FILENAME

Generate a patch:
\fBkatana\fP [-c CONFIG]-g [-o OUTPUT_FILE] OLD_OBJECTS_DIR NEW_OBJECTS_DIR EXECUTABLE_NAME

Apply a patch:
\fBkatana\fP [-c CONFIG] -p [-s] PATCH_FILENAME PID

List patch information:
\fBkatana\fP -l PATCH_FILE


.\ enable filling again
.fi



.SH OVERVIEW
Katana was developed initially for hotpatching. It was then extended
as a more general binary manipulation framework, especially for
research into security issues surrounding the DWARF format. The rest
of this manual page deals with use for hotpatching. For information on
the usage of the Katana shell, see the complete Katana User's Guide which
can be found at http://katana.nongnu.org or in the katana subdirectory
of your system's standard location for installing documentation.

As a patcher, Katana is intended to be used in two stages. The first
stage generates a patch object from two different versions of an
treee. By an object tree, we mean the set of object files (.o files)
and the executable binary they comprise. Katana works completely at
the object level, so the source code itself is not strictly required,
although all objects must be compiled with debugging information. This
step may be done by the software vendor. In the second stage, the
patch is applied to a running process. The original source trees are
not necessary during patch application, as the patch object contains
all information necessary to patch the in-memory process at the object
level. It is also possible to view the contents of a patch object in a
human-readable way for the purposes of sanity-checking, determining
what changes the patch makes, etc.

.SH Preparing a Package for Patching Support
Katana aims to be much less invasive than other hot-patching system
and require minimal work to be used with any project. It does,
however, have some requirements.

Required CFLAGS:
  -g

Recommended CFLAGS:
  -ffunction-sections
  -fdata-sections
     
Recommended LDFLAGS:
  --emit-relocs

.SH GENERATING A PATCH
Let the location of your project be /project. You must have two
versions of your software available: the version identical to the
running software which must be hotpatched, call it v0, and the version
to which you wish to hotpatch the running software, call it v1. Let
foo be the name of your program. Then /project/v0/foo must exist and
/project/v0 must also contain (possibly in subdirectories) all of the
object files which contributed to /project/v0/foo. The source code
itself is immaterial, as Katana does not parse it. Similarly,
/project/v1/foo must exist and /project/v1 contain all of the object
files contributing to /project/v1/foo. Katana is then invoked as

\fBkatana\fP -g [-o OUTPUT_FILE] /project/v0 /project/v1 foo

If -o OUTPUT_FILE is not specified, the output file will be OLD_OBJECTS_DIR/EXECUTABLE_NAME.po
(in this case /project/v0/foo.po)

.SH APPLYING A PATCH
The process to be patched is running with a pid of PID. It can be
patched from its current version to a more recent version by the Patch
Object (PO) file PATCH. Katana is then invoked as

\fBkatana\fP -p [-s] PATCH_FILENAME PID

If all goes well, the patcher will run, print out some status
messages, and leave your program in better state than it found it. The
optional -s flag tells Katana to stop the target program after
patching it and detaching from it. This is mostly of use for debugging
Katana.

Katana requires permission to ptrace the target process. This can
always be performed by running Katana as root. When that is not
desirable but ptrace is limited (e.g. by Yama LSM), Katana can be
granted CAP_PTRACE by running

setcap cap_sys_ptrace=eip /path/to/katana

.SH VIEWING A PATCH
One of the goals of Katana and its Patch Object (PO) format is to
increase the transparency of patches: a user about to apply a patch
should know what it will do. This goal is not yet fully realized, but
it is possible to view some information about a patch with

\fBkatana\fP -l PATCH_FILENAME

.SH CONFIGURATION FILES
The -c CONFIG option tells katana to load the configuration file
CONFIG. Additionally, and of the files /etc/katana, ~/.katana,
~/.config/katana, and ./.katana which exist will be loaded (in that
order) as configuration files regardless of the -c option. The CONFIG
file specified by -c will be loaded last. Later configuration files
may override settings from earlier ones. Configuration files are
written in JSON (with the requirement to quote strings relaxed). For a
complete guide to configuration options, see the Katana User's Guide
which should be available at http://katana.nongnu.org.

.SH SEE ALSO
Read the complete Katana User's Guide which should be available in
several formats at http://katana.nongnu.org or in the katana
subdirectory of your system's standard location for installing
documentation.

.SH AUTHORS
Katana was written by James Oakley (<james.oakley@dartmouth.edu> or
<electron100@gmail.com>). It was designed by Sergey Bratus, James
Oakley, Ashwin Ramaswamy, Michael E. Locasto, and Sean
W. Smith. Portions of Katana are Copyright (C) 2010-2011 Dartmouth
College.

Portions of this document are Copyright 2010-2011 Dartmouth College.
This document may be distributed under the terms of the GNU Free
Documentation License as found in the file FDL which should have been
distributed with this software. If it was not, it may be found at
http://www.gnu.org/licenses/fdl.txt.


ChangeLog
2011-04-13  james  <james.oakley@dartmouth.edu>

	* src/callFrameInfo.c fixed alignment issue for the type table

	* src/elfwriter.c flagging data sections with SHT_NOBITS as dirty was
	causing bizarre crashes on x86. Didn't fully debug, but simply
	don't flag them as dirty (since there's not really any
	need afaik). Should perhaps eventually give them more attention.

2011-04-13  James Oakley  <james.oakley@dartmouth.edu>
	* katana finally has a ChangeLog



src/katana.c
/*
  File: katana.c
  Author: James Oakley
  Copyright (C): 2010 Dartmouth College
  License: Katana is free software: you may redistribute it and/or
    modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation, either version 2 of the
    License, or (at your option) any later version. Regardless of
    which version is chose, the following stipulation also applies:
    
    Any redistribution must include copyright notice attribution to
    Dartmouth College as well as the Warranty Disclaimer below, as well as
    this list of conditions in any related documentation and, if feasible,
    on the redistributed software; Any redistribution must include the
    acknowledgment, “This product includes software developed by Dartmouth
    College,” in any related documentation and, if feasible, in the
    redistributed software; and The names “Dartmouth” and “Dartmouth
    College” may not be used to endorse or promote products derived from
    this software.  

                             WARRANTY DISCLAIMER

    PLEASE BE ADVISED THAT THERE IS NO WARRANTY PROVIDED WITH THIS
    SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING, DARTMOUTH COLLEGE, ANY OTHER COPYRIGHT
    HOLDERS, AND/OR OTHER PARTIES PROVIDING OR DISTRIBUTING THE SOFTWARE,
    DO SO ON AN "AS IS" BASIS, WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
    PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE
    SOFTWARE FALLS UPON THE USER OF THE SOFTWARE. SHOULD THE SOFTWARE
    PROVE DEFECTIVE, YOU (AS THE USER OR REDISTRIBUTOR) ASSUME ALL COSTS
    OF ALL NECESSARY SERVICING, REPAIR OR CORRECTIONS.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL DARTMOUTH COLLEGE OR ANY OTHER COPYRIGHT HOLDER, OR ANY OTHER
    PARTY WHO MAY MODIFY AND/OR REDISTRIBUTE THE SOFTWARE AS PERMITTED
    ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL,
    INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR
    INABILITY TO USE THE SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF
    DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR
    THIRD PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER
    PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
    POSSIBILITY OF SUCH DAMAGES.

  
    The complete text of the license may be found in the file COPYING
    which should have been distributed with this software. The GNU
    General Public License may be obtained at
    http://www.gnu.org/licenses/gpl.html

  Project: Katana
  Date: December, 2010
  Description:
*/
#include <errno.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <assert.h>
#include <signal.h>
#include <sys/wait.h>
#include "dwarftypes.h"
#include "types.h"
#include "elfparse.h"
#include <signal.h>
#include "patchwrite/patchwrite.h"
ElfInfo* oldBinElfInfo=NULL;
#ifdef DEBUG
#include "katana-prelim.h"
#endif
#include <sys/stat.h>
#include "patcher/patchapply.h"
#include "patcher/versioning.h"
#include "util/logging.h"
#include "patchwrite/typediff.h"
#include "info/fdedump.h"
#include "info/dwinfo_dump.h"
#include "info/unsafe_funcs_dump.h"
#include "util/path.h"
#include "katana_config.h"
#include "rewriter/rewrite.h"
#include "shell/shell.h"

void configureFromCommandLine(int argc,char** argv);

void sigsegReceived(int signum)
{
  death("katana segfaulting. . .\n");
}

int main(int argc,char** argv)
{
  #ifdef DEBUG
  system("rm -rf /tmp/katana-$USER/patched/*");
  #endif
  loggingDefaults();
  setDefaultConfig();

  struct stat s;
  if(0==stat("/etc/katana",&s))
  {
    loadConfigurationFile("~/.katana");
  }
  if(0==stat("~/.katana",&s))
  {
    loadConfigurationFile("~/.katana");
  }
  if(0==stat("~/.config/katana",&s))
  {
    loadConfigurationFile("~/.config/katana");
  }
  if(0==stat(".katana",&s))
  {
    loadConfigurationFile(".katana");
  }
  
  struct sigaction act;
  memset(&act,0,sizeof(struct sigaction));
  act.sa_handler=&sigsegReceived;
  sigaction(SIGSEGV,&act,NULL);
  if(elf_version(EV_CURRENT)==EV_NONE)
  {
    death("Failed to init ELF library\n");
  }
  configureFromCommandLine(argc,argv);
  if(EKM_SHELL==config.mode)
  {
    doShell(config.inputFile);
  }
  else if(EKM_GEN_PATCH==config.mode)
  {
    char* oldBinPath=joinPaths(config.oldSourceTree,config.objectName);
    char* newBinPath=joinPaths(config.newSourceTree,config.objectName);
    
    if(!config.outfileName)
    {
      config.outfileName=zmalloc(strlen(oldBinPath)+5);
      strcpy(config.outfileName,oldBinPath);
      strcat(config.outfileName,".po");
    }

    ElfInfo* patch=createPatch(config.oldSourceTree,config.newSourceTree,oldBinPath,newBinPath,NULL,config.outfileName);
    endELF(patch);
    free(oldBinPath);
    free(newBinPath);
  }
  else if(EKM_APPLY_PATCH==config.mode)
  {
    oldBinElfInfo=getElfRepresentingProc(config.pid);
    findELFSections(oldBinElfInfo);
    ElfInfo* patch=openELFFile(config.objectName);
    findELFSections(patch);
    patch->isPO=true;
    readAndApplyPatch(config.pid,oldBinElfInfo,patch);
    endELF(patch);
  }
  else if(EKM_INFO==config.mode)
  {
    ElfInfo* patch=openELFFile(config.objectName);
    Map* fdeMap=readDebugFrame(patch,isFlag(EKCF_EH_FRAME));
    if(!fdeMap)
    {
      death("Unable to read frame info, can't print info\n");
    }
    printf("*********Type and Function Info****************\n");
    printPatchDwarfInfo(patch,fdeMap);

    //while katana is primarily for dealing with patches, it also has
    //some functionality for displaying general-purpose information
    //about DWARF and ELF information.
    if(patch->isPO)
    {
      printf("\n*********Safety Info****************\n");
      printPatchUnsafeFuncsInfo(patch);
    }
    if(patch->isPO)
    {
      printf("\n*********Type Transformation Rules****************\n");
    }
    else
    {
      printf("\n*********Call Frame Information****************\n");
    }
    printCallFrameInfo(stdout,&patch->callFrameInfo,patch);
  }
  else if(EKM_TEST_PASSTHROUGH==config.mode)
  {
    ElfInfo* object=openELFFile(config.objectName);
    rewrite(object,config.outfileName);
  }
  else
  {
    death("unhandled katana mode");
  }
  return 0;
}

